components/salsa-2022/src/input_field.rs

```diff
@@ -0,0 +1,67 @@
+use std::hash::Hash;
+use rustc_hash::FxHashMap;
+use crate::{Durability, IngredientIndex, Revision, Runtime};
+use crate::cycle::CycleRecoveryStrategy;
+use crate::ingredient::Ingredient;
+use crate::key::DependencyIndex;
+use crate::runtime::local_state::QueryInputs;
+use crate::runtime::StampedValue;
+
+pub struct InputFieldIngredient<K, F> {
+    index: IngredientIndex,
+    map: FxHashMap<K, StampedValue<F>>
+}
+
+impl<K, F> InputFieldIngredient<K, F>
+where K: Eq + Hash
+{
+    pub fn new(index: IngredientIndex) -> Self {
+        Self {
+            index,
+            map: Default::default(),
+        }
+    }
+
+    pub fn store(
+        &mut self,
+        runtime: &mut Runtime,
+        key: K,
+        value: F,
+        durability: Durability,
+    ) -> Option<F> {
+        let revision = runtime.current_revision();
+        let stamped_value = StampedValue {
+            value,
+            durability,
+            changed_at: revision
+        };
+
+        if let Some(old_value) = self.map.insert(key, stamped_value) {
+            Some(old_value.value)
+        } else {
+            None
+        }
+    }
+
+    pub fn fetch(
+        &self,
+        key: K,
+    ) -> &F {
+        &self.map.get(&key).unwrap().value
+    }
+}
+
+impl<DB: ?Sized, K, F> Ingredient<DB> for InputFieldIngredient<K, F>
+{
+    fn cycle_recovery_strategy(&self) -> CycleRecoveryStrategy {
+        CycleRecoveryStrategy::Panic
+    }
+
+    fn maybe_changed_after(&self, _db: &DB, _input: DependencyIndex, _revision: Revision) -> bool {
```

Comment:
This needs to 

First, convert `input` into a `K` by doing something like this:

https://github.com/salsa-rs/salsa/blob/8b7324dca886001df7e7bf7c45567520954fec6b/components/salsa-2022/src/function.rs#L193

Then, lookup the `StampedValue` in `self.map` and compare `_revision` against the `changed_at` field, something like this:

```rust
self.map[&input].changed_at > revision
```

similar to this:

https://github.com/salsa-rs/salsa/blob/master/components/salsa-2022/src/function/maybe_changed_after.rs#L40

except that the `memo` is just the `StampedValue` we read from `self.map`

Reply:
In https://github.com/salsa-rs/salsa/pull/342/commits/740fe86fe98c4ab8cf4bdfce752d7b70b0ea191e I implemented something similar to what `FunctionIngredient` has, but without adding the `key_from_id` helper function.